### ISSUE TITLE
wrong "fruitless type test" warning depending on source file order

### ISSUE THREAD

**Issue by lrytz** [scala/bug#12703]
Scala 2.13.10

```
➜ sandbox cat A.scala
object Foo {
  sealed trait A
  sealed trait B
  final case class C(x: Int) extends A with B
}

➜ sandbox cat B.scala
object T {
  import Foo._
  def f(a: A) = a match {
    case b: B => 0
    case _ => 1
  }
}

➜ sandbox sc A.scala B.scala

➜ sandbox sc B.scala A.scala
B.scala:4: warning: fruitless type test: a value of type Foo.A cannot also be a Foo.B
    case b: B => 0
            ^
1 warning
```

**Comment by lrytz** [comment]
The check is executed here when type checking the first source file finishes: https://github.com/scala/scala/blob/v2.13.10/src/compiler/scala/tools/nsc/typechecker/Analyzer.scala#L120

The lazy type of the second source file is not yet complete, the children will only be populated when it's forced in type checking.

**Comment by som-snytt** [comment]
No idea if this is related to https://github.com/scala/bug/issues/12438 and its proposed fix. It occurs to me that some checks benefit from coming later, assuming no loss of information. In phase travel, hindsight is 20/20.

**Comment by lrytz** [comment]
We used to skip the "fruitless" test for symbols compiled in the current run (leading to missing warnings). https://github.com/scala/scala/pull/9668 fixed that, but the deferred check is still too early, it should run only after all compilation units went through typer.

**Comment by lrytz** [comment]
Maybe all of `checkCheckable` can be deferred to RefChecks, it doesn't seem to affect type checking / translation.

Prior art: [`TypeTreeWithDeferredRefCheck`](https://github.com/scala/scala/blob/v2.13.10/src/compiler/scala/tools/nsc/ast/Trees.scala#L69)
